ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java

```diff
@@ -0,0 +1,260 @@
+   * The output schema comprises of exactly the columns that appear in the SELECT clause of the
+   * query.
+   * @param metaStore the metastore
+   * @return the project node's output schema
+   */
+  @VisibleForTesting
+  protected LogicalSchema buildOutputSchema(final MetaStore metaStore) {
+    final LogicalSchema outputSchema;
+    final LogicalSchema parentSchema = getSource().getSchema();
+    final boolean isWindowed = analysis
+        .getFrom()
+        .getDataSource()
+        .getKsqlTopic()
+        .getKeyFormat().isWindowed();
+
+    if (isSelectStar()) {
+      outputSchema = buildPullQuerySelectStarSchema(
+          parentSchema.withoutPseudoAndKeyColsInValue(), isWindowed);
+    } else {
+      final List<SelectExpression> projects = projection.selectItems().stream()
+          .map(SingleColumn.class::cast)
+          .map(si -> SelectExpression
+              .of(si.getAlias().orElseThrow(IllegalStateException::new), si.getExpression()))
+          .collect(Collectors.toList());
+
+      outputSchema = selectOutputSchema(metaStore, projects, isWindowed);
+    }
+    return outputSchema;
+  }
+
+  /**
+   * Checks whether the intermediate schema should be extended with system and key columns.
+   * @return true if the intermediate schema should be extended
+   */
+  private boolean shouldAddAdditionalColumnsInSchema() {
+
+    final boolean hasSystemColumns = analysis.getSelectColumnNames().stream().anyMatch(
+        SystemColumns::isSystemColumn
+    );
+
+    final boolean hasKeyColumns = analysis.getSelectColumnNames().stream().anyMatch(cn ->
+        getSource().getSchema().isKeyColumn(cn)
+    );
+
+    return hasSystemColumns || hasKeyColumns;
+  }
+
+  private boolean isSelectStar() {
+    final boolean someStars = projection.selectItems().stream()
+        .anyMatch(s -> s instanceof AllColumns);
+
+    if (someStars && projection.selectItems().size() != 1) {
+      throw new KsqlException("Pull queries only support wildcards in the projects "
+                                  + "if they are the only expression");
+    }
+
+    return someStars;
+  }
+
+  private LogicalSchema buildPullQuerySelectStarSchema(
+      final LogicalSchema schema,
+      final boolean windowed
+  ) {
+    final Builder builder = LogicalSchema.builder()
+        .keyColumns(schema.key());
+
+    if (windowed) {
+      builder.keyColumn(SystemColumns.WINDOWSTART_NAME, SqlTypes.BIGINT);
+      builder.keyColumn(SystemColumns.WINDOWEND_NAME, SqlTypes.BIGINT);
+    }
+
+    return builder
+        .valueColumns(schema.value())
+        .build();
+  }
+
+  private LogicalSchema selectOutputSchema(
+      final MetaStore metaStore,
+      final List<SelectExpression> selectExpressions,
+      final boolean isWindowed
+  ) {
+    final Builder schemaBuilder = LogicalSchema.builder();
+    final LogicalSchema parentSchema = getSource().getSchema();
+
+    // Copy meta & key columns into the value schema as SelectValueMapper expects it:
```

Comment:
Does this mean ExpressionTypeManager expects it?  That's the main place where this schema is used is just to evaluate the type of the expression.

Reply:
Also, the `CodeGenRunner` uses the schema

Reply:
I guess I just meant that `schema` is only used with the `ExpressionTypeManager` and the comment is a bit misleading.